python/python/tests/test_async_client.py

```diff
@@ -7303,6 +7303,365 @@ async def test_lcs_idx(self, redis_client: GlideClient):
+        key = get_random_string(10)
+        expected_keys = [f"key:{key}:{i}" for i in range(100)]
+        await redis_client.mset({k: "value" for k in expected_keys})
+        cursor = ClusterScanCursor()
+        keys: list[str] = []
+        while not cursor.is_finished():
+            result = await redis_client.scan(cursor, match="key:*")
+            cursor = cast(ClusterScanCursor, result[0])
+            result_keys = cast(list[str], result[1])
+            keys.extend(result_keys)
+        assert set(expected_keys) == set(keys)
+        assert not set(unexpected_keys).intersection(set(keys))
+
+    @pytest.mark.parametrize("cluster_mode", [True])
+    @pytest.mark.parametrize("protocol", [ProtocolVersion.RESP2, ProtocolVersion.RESP3])
+    # We test whether the cursor is cleaned up after it is deleted, which we expect to happen when th GC is called
+    async def test_cluster_scan_cleaning_cursor(self, redis_client: GlideClusterClient):
+        key = get_random_string(10)
+        await redis_client.mset(
+            {k: "value" for k in [f"{key}:{i}" for i in range(100)]}
+        )
+        cursor = ClusterScanCursor()
+        result = await redis_client.scan(cursor)
+        cursor = cast(ClusterScanCursor, result[0])
+        cursor_string = str(cursor)
+        del cursor
+        new_cursor_with_same_id = ClusterScanCursor(cursor_string)
+        with pytest.raises(RequestError) as e_info:
+            await redis_client.scan(new_cursor_with_same_id)
+        assert "Invalid scan_state_cursor id" in str(e_info.value)
+
+    @pytest.mark.parametrize("cluster_mode", [True])
+    @pytest.mark.parametrize("protocol", [ProtocolVersion.RESP2, ProtocolVersion.RESP3])
+    async def test_cluster_scan_all_types(self, redis_client: GlideClusterClient):
+        # We test that the scan command work for all types of keys
+        key = get_random_string(10)
+        string_keys = [f"{key}:{i}" for i in range(100)]
+        await redis_client.mset({k: "value" for k in string_keys})
+
+        set_key = get_random_string(10)
+        set_keys = [f"{set_key}:{i}" for i in range(100, 200)]
+        for key in set_keys:
+            await redis_client.sadd(key, ["value"])
+
+        hash_key = get_random_string(10)
+        hash_keys = [f"{hash_key}:{i}" for i in range(200, 300)]
+        for key in hash_keys:
+            await redis_client.hset(key, {"field": "value"})
+
+        list_key = get_random_string(10)
+        list_keys = [f"{list_key}:{i}" for i in range(300, 400)]
+        for key in list_keys:
+            await redis_client.lpush(key, ["value"])
+
+        zset_key = get_random_string(10)
+        zset_keys = [f"{zset_key}:{i}" for i in range(400, 500)]
+        for key in zset_keys:
+            await redis_client.zadd(key, {"value": 1})
+
+        cursor = ClusterScanCursor()
+        keys: list[str] = []
+        while not cursor.is_finished():
+            result = await redis_client.scan(cursor, type=ObjectType.STRING)
```

Comment:
Should the old cursor get cleaned up in between scan calls?

Reply:
Nope, as mentioned in previous comments. The state of the scan is not managed for the user other than creating a new object for each iteration.
Every cursor is pointing to a different object so we do not need to clean it. It will be cleaned by the gc when the user will drop its reference. If the user wants to hold some hold cursor for whatever wired reasons, its fine, he can. I don't decide for him when to drop a cursor. 
As every other object in languages with gc, reference count is what decides the lifetime of the objects.
When you create a list in python you manually delete it when you don't use it anymore, you trust that the language will do it, that's the all idea of languages without memory management. You pay in efficiency, but you don't need to manage manually.
But the important part is that the new cursor is not pointing to the same state object (the internal one on redia-rs).
Its a stateless function which rely only on your input (the cursor you sent) and the state of the db at the moment. Not on something that we keep updating and saving for you.

Reply:
While you don't manually de-allocate memory in Python, you do use with statements to control the lifetime of other limited resources.

Even in Java, you don't manually release JVM memory, but you do have to manually release direct memory that gets allocated. Generally anything that does resource clean-up in a finalizer is paired with a close method for deterministic clean-up to let the user ensure there's no resource exhaustion.